cmd/vega/propose_upgrade.go

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2022 Gobalsky Labs Limited
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at https://www.mariadb.com/bsl11.
+//
+// Change Date: 18 months from the later of the date of the first publicly
+// available Distribution of this version of the repository, and 25 June 2022.
+//
+// On the date above, in accordance with the Business Source License, use
+// of this software will be governed by version 3 or later of the GNU General
+// Public License.
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	commandspb "code.vegaprotocol.io/protos/vega/commands/v1"
+	vgcrypto "code.vegaprotocol.io/shared/libs/crypto"
+	"code.vegaprotocol.io/shared/paths"
+	"code.vegaprotocol.io/vega/config"
+	"code.vegaprotocol.io/vega/logging"
+	"code.vegaprotocol.io/vega/txn"
+
+	"github.com/jessevdk/go-flags"
+)
+
+type ProposeUpgradeCmd struct {
+	config.VegaHomeFlag
+	config.OutputFlag
+	config.Passphrase `long:"passphrase-file"`
+
+	VegaReleaseTag     string `short:"v" long:"vega-release-tag" required:"true" description:"A valid vega core release tag for the upgrade proposal"`
+	DatanodeReleaseTag string `short:"d" long:"datanode-release-tag" required:"true" description:"A valid data-node release tag for upgrade proposal"`
+	UpgradeBlockHeight uint64 `short:"h" long:"height" required:"true" description:"The block height at which the upgrade should be made"`
+}
+
+var proposeUpgradeCmd ProposeUpgradeCmd
+
+func (opts *ProposeUpgradeCmd) Execute(args []string) error {
+	log := logging.NewLoggerFromConfig(logging.NewDefaultConfig())
+	defer log.AtExit()
+
+	registryPass, err := opts.Get("node wallet", false)
+	if err != nil {
+		return err
+	}
+
+	vegaPaths := paths.New(opts.VegaHome)
+
+	_, conf, err := config.EnsureNodeConfig(vegaPaths)
+	if err != nil {
+		return err
+	}
+
+	if !conf.IsValidator() {
+		return errors.New("node is not a validator")
+	}
+
+	cmd := commandspb.ProtocolUpgradeProposal{
+		VegaReleaseTag:     opts.VegaReleaseTag,
+		DataNodeReleaseTag: opts.DatanodeReleaseTag,
+		UpgradeBlockHeight: opts.UpgradeBlockHeight,
```

Comment:
Should we add validation of the semver here / check if the block is non-zero but also is a future block? Just so we can add nice error handling here.

Reply:
added